src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py

```diff
@@ -0,0 +1,410 @@
+        optimize_mode : str
+            optimize mode, 'maximize' or 'minimize', by default 'maximize'
+        pruning_mode : str
+            'channel' or 'fine_grained', by default 'channel'
+        experiment_data_dir : str
+            PATH to save experiment data
+        """
+        # models used for iterative pruning and evaluation
+        self._model_to_prune = copy.deepcopy(model)
+        self._pruning_mode = pruning_mode
+
+        super().__init__(model, config_list)
+
+        self._fine_tuner = fine_tuner
+        self._evaluator = evaluator
+        self._optimize_mode = OptimizeMode(optimize_mode)
+
+        # hyper parameters for NetAdapt algorithm
+        self._pruning_step = pruning_step
+
+        # overall pruning rate
+        self._sparsity = config_list[0]['sparsity']
+
+        # config_list
+        self._config_list = []
+
+        self._experiment_data_dir = experiment_data_dir
+        if not os.path.exists(self._experiment_data_dir):
+            os.makedirs(self._experiment_data_dir)
+
+        self._tmp_model_path = os.path.join(
+            self._experiment_data_dir, 'tmp_model.pth')
+
+        self._total_num_weights = self._get_total_num_weights()
+
+    def _detect_modules_to_compress(self):
+        """
+        redefine this function, consider only the layers without dependencies
+        """
+        self._op_names_to_compress = []
+        if self.modules_to_compress is None:
+            self.modules_to_compress = []
+            # consider only the layers without dependencies
+            model_name = self._model_to_prune.__class__.__name__
+            ops_no_dependency = get_layers_no_dependency(model_name)
+
+            for name, module in self.bound_model.named_modules():
+                if module == self.bound_model:
+                    continue
+                if self._pruning_mode == 'channel' and model_name in ['MobileNetV2', 'RetinaFace'] and name not in ops_no_dependency:
+                    continue
+                layer = LayerInfo(name, module)
+                config = self.select_config(layer)
+                if config is not None:
+                    self.modules_to_compress.append((layer, config))
+                    self._op_names_to_compress.append(name)
+        return self.modules_to_compress
+
+    def _get_total_num_weights(self, count_non_prunable_ops=True):
+        '''
+        calculate the total number of weights
+
+        Parameters
+        ----------
+        count_non_prunable_ops : bool
+            indicate if non prunable ops should be considerd
+
+        Returns
+        -------
+        int
+            total weights of all the op considered
+        '''
+        num_weights = 0
+        if count_non_prunable_ops:
+            pruner = LevelPruner(
+                copy.deepcopy(self._model_to_prune), [{'sparsity': 0.1, 'op_types': ['default']}])
+            modules_wrapper = pruner.get_modules_wrapper()
```

Comment:
does the deepcopy support all models?

Reply:
Yes, this is a solution recommended in the PyTorch forum to copy a model.